model/value.go

```diff
@@ -129,53 +59,89 @@ func (s *Sample) Equal(o *Sample) bool {
 	if !s.Timestamp.Equal(o.Timestamp) {
 		return false
 	}
-
-	return s.Value.Equal(o.Value)
+	if s.Type != o.Type {
+		return false
+	}
+	if s.Type == STFloat {
+		return s.Value.Equal(o.Value)
+	}
+	return s.Histogram.Equal(&o.Histogram)
 }
 
 func (s Sample) String() string {
-	return fmt.Sprintf("%s => %s", s.Metric, SamplePair{
-		Timestamp: s.Timestamp,
-		Value:     s.Value,
-	})
+	if s.Type == STFloat {
+		return fmt.Sprintf("%s => %s", s.Metric, SamplePair{
+			Timestamp: s.Timestamp,
+			Value:     s.Value,
+		})
+	} else {
+		return fmt.Sprintf("%s => %s", s.Metric, SampleHistogramPair{
+			Timestamp: s.Timestamp,
+			Histogram: s.Histogram,
+		})
+	}
 }
 
 // MarshalJSON implements json.Marshaler.
 func (s Sample) MarshalJSON() ([]byte, error) {
-	v := struct {
-		Metric Metric     `json:"metric"`
-		Value  SamplePair `json:"value"`
-	}{
-		Metric: s.Metric,
-		Value: SamplePair{
-			Timestamp: s.Timestamp,
-			Value:     s.Value,
-		},
+	if s.Type == STFloat {
+		v := struct {
+			Metric Metric     `json:"metric"`
+			Value  SamplePair `json:"value"`
+		}{
+			Metric: s.Metric,
+			Value: SamplePair{
+				Timestamp: s.Timestamp,
+				Value:     s.Value,
+			},
+		}
+		return json.Marshal(&v)
+	} else {
+		v := struct {
+			Metric    Metric              `json:"metric"`
+			Histogram SampleHistogramPair `json:"histogram"`
+		}{
+			Metric: s.Metric,
+			Histogram: SampleHistogramPair{
+				Timestamp: s.Timestamp,
+				Histogram: s.Histogram,
+			},
+		}
+		return json.Marshal(&v)
 	}
-
-	return json.Marshal(&v)
 }
 
 // UnmarshalJSON implements json.Unmarshaler.
 func (s *Sample) UnmarshalJSON(b []byte) error {
 	v := struct {
-		Metric Metric     `json:"metric"`
-		Value  SamplePair `json:"value"`
+		Metric    Metric              `json:"metric"`
+		Value     SamplePair          `json:"value"`
+		Histogram SampleHistogramPair `json:"histogram"`
 	}{
 		Metric: s.Metric,
 		Value: SamplePair{
 			Timestamp: s.Timestamp,
 			Value:     s.Value,
 		},
+		Histogram: SampleHistogramPair{
+			Timestamp: s.Timestamp,
+			Histogram: s.Histogram,
+		},
 	}
 
 	if err := json.Unmarshal(b, &v); err != nil {
 		return err
 	}
 
 	s.Metric = v.Metric
-	s.Timestamp = v.Value.Timestamp
-	s.Value = v.Value.Value
+	if v.Histogram.Timestamp != 0 {
```

Comment:
Fixed this, but did something very hacky with a local struct called `sampleHistogramPairPtr` because I don't know how else to fix the unmarshalling with a pointer.